### PR TITLE
nixos/nginx: use acme locations only on port 80

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
ACME only uses port 80 to obtain certificates. More information - https://letsencrypt.org/docs/challenge-types/#http-01-challenge

cc @NixOS/acme
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by m1cr0man** [comment]
@ofborg test acme

**Review by m1cr0man — APPROVED** [review]
Thanks for submitting this. It seems harmless enough, and the test suite is happy.

**Comment by ncfavier** [comment]
This leaves `acmeLocation` only in the redirection server, which is not included if `forceSSL` is false. Is this intended?

**Comment by m1cr0man** [comment]
Yeah, as pointed out it's only needed on port 80/HTTP anyway for HTTP-01 validation. It never needed to be in the HTTPS server. I think I didn't remove it myself in abundance of caution

**Comment by ncfavier** [comment]
To be clear, unless I'm misunderstanding something, this will **break** scenarios with `forceSSL = false;` as the `acmeLocation` won't be included in any servers. An easy fix would be to keep the `acmeLocation` in the `hostListen` server if `forceSSL` is set to false.

I'm also not sure how to interpret those two sentences from the link above:

> It only accepts redirects to “http:” or “https:”, and only to ports 80 or 443.

> The HTTP-01 challenge can only be done on port 80.

Don't they contradict each other?

**Comment by m1cr0man** [comment]
HTTP-01 validation requires a server listening on port 80 to serve the acmeLocation to LetsEncrypt. If you `forceSSL = false`, the assumption is that you aren't serving on port 80 at all, and thus you can't use HTTP-01 validation. This is where the confusion lies I believe - looking over the Nginx module I suppose it's possible to `forceSSL = false` and still have a port 80 listener (via the `listen` option on the vhost submodule).

In an abundance of caution we could keep the acmeLocation where it is to cover this scenario. However, @izorkin, does this PR fix an issue that you encountered or was it only for the sake of clean up?

**Comment by ncfavier** [comment]
An abundance of caution? I'm only talking about **not breaking existing setups** (those using `forceSSL = false;`).

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -307,7 +307,6 @@ let
         server {
           ${concatMapStringsSep "\n" listenString hostListen}
           server_name ${vhost.serverName} ${concatStringsSep " " vhost.serverAliases};
-          ${acmeLocation}
           ${optionalString (vhost.root != null) "root ${vhost.root};"}
```
Comment (ncfavier) [inline nixos/modules/services/web-servers/nginx/default.nix:310]:
```suggestion
          ${optionalString (!vhost.forceSSL) acmeLocation}
          ${optionalString (vhost.root != null) "root ${vhost.root};"}
```

**Review by ncfavier — CHANGES_REQUESTED** [review]

**Comment by m1cr0man** [comment]
Oh of course, I suppose you can configure a vhost with port 80 and port 443 with no SSL enforcement 😅 that didn't even come to mind as a valid configuration since I've been so used to HTTPS at this point. Yeah, this needs to stay. Sorry for the confusion.

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (m1cr0man) [reply nixos/modules/services/web-servers/nginx/default.nix:310]: Lets not even complicate it further just to support this case. The PR should probably be dropped as a whole.

Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:310]: Thanks, fixed.

**Comment by Izorkin** [comment]
Update PR.

**Comment by Izorkin** [comment]
> In an abundance of caution we could keep the acmeLocation where it is to cover this scenario. However, @Izorkin, does this PR fix an issue that you encountered or was it only for the sake of clean up?

Tried to cleanup configuration.

**Comment by ncfavier** [comment]
To be fair I'm not against adding the assertion, but I think a better test would be to check that `defaultListen` has at least one address with port 80.

I'd really like to wait for a second opinion from @NixOS/acme though since I still don't understand how HTTP-01 works.

**Comment by Izorkin** [comment]
> This ends up _adding_ code, so probably not worth it.

Reduced code in nginx.conf.

**Comment by m1cr0man** [comment]
I want to wrap this up instead of leaving the PR dangling. There's a balance that has to be reached between complexity in the Nix module and in the target application config. This has been prevalent in the ACME module for quite some time, where we keep adding complexity to handle edge cases that otherwise can't be fixed, and the module continues to get harder to maintain.

In the case of this PR, having acmeLocation in the Nginx config unconditionally does not break anything. For the sake of maintainability, I'd rather not add these extra conditionals to the module. There's no issue being fixed here other than simplifying the configuration, which I don't think is a strong enough motive for a merge here.

If we're in agreement, I want to close this PR. I do appreciate you taking the time to make the PR, but I don't think it's something we should change.

**Comment by Izorkin** [comment]
After several months of using this changes I did not find any problems with certificates :)

**Comment by m1cr0man** [comment]
Glad to hear it's working Izorkin, however I am still not convinced this needs to be merged. I am going to close this PR now but if there is a technical problem this ends up solving we can revisit it in the future.

**Comment by Izorkin** [comment]
@m1cr0man allof this simple variant?
```diff
diff --git a/nixos/modules/services/web-servers/nginx/default.nix b/nixos/modules/services/web-servers/nginx/default.nix
index bc5b4be5109..3881ec4c12f 100644
--- a/nixos/modules/services/web-servers/nginx/default.nix
+++ b/nixos/modules/services/web-servers/nginx/default.nix
@@ -357,7 +357,7 @@ let
         server {
           ${concatMapStringsSep "\n" listenString hostListen}
           server_name ${vhost.serverName} ${concatStringsSep " " vhost.serverAliases};
-          ${acmeLocation}
+          ${optionalString ((!vhost.forceSSL && !hasSSL) || (!vhost.forceSSL && !onlySSL)) acmeLocation}
           ${optionalString (vhost.root != null) "root ${vhost.root};"}
           ${optionalString (vhost.globalRedirect != null) ''
             location / {

```

**Comment by m1cr0man** [comment]
I know it seems petty but these sort of little changes add up. Small changes are how the ACME module has become the monster it is right now. I'll be more specific about my problems:

- acmeLocation is [already wrapped in conditionals](https://github.com/NixOS/nixpkgs/blob/532f202b6490c646817796ba09e88c8b7bea74c1/nixos/modules/services/web-servers/nginx/default.nix#L327) to generate nothing in certain circumstances
- There is no problem having the listener + locations on port 80 right now
- Adding extra logic around acmeLocation complicates future refactoring + diagnosing
- The assertion is not actually valid since it is possible to reconfigure the default ACME solving method to use DNS or other mechanisms and avoid the need for a HTTP-01 server in nginx.

As for the smaller change in your last message, that set of conditions is quite confusing with all the negations but I built a truth table to make sense of it:

|onlySSL|forceSSL|hasSSL|!force+!has|!force+!only|explain|
|-|-|-|-|-|-|
|T|F|T|F|F|Only listen on port 443|
|F|T|T|F|F|Force redirect to 443|
|T|T|T|F|F|Only listen 443 + force redirect (impossible combo)|
|F|F|T|F|T|addSSL or enableSSL are true|
|F|F|F|T|T|plain HTTP server|

hasSSL will be true if onlySSL or forceSSL are true, as well as if addSSL or enableSSL are true. Having a conditional pair between forceSSL and hasSSL doesn't make sense.

Alas, this took me 20 minutes to figure out. I don't want regular users to have to go this deep into the nginx SSL configuration to figure out why their acmeLocation stanzas are disappeared and when they should render.

**Comment by Izorkin** [comment]
I manually tested different combinations without making a table of results. I missed the simpler variant.
Ok, let's leave it at that.

**Pull request closed without merging**